### ISSUE TITLE
race condition in ThreadChildWatcher (default) and MultiLoopChildWatcher

### ISSUE THREAD

**Issue by b260794d-9f59-47d2-a76e-b129bedbca29** [python/cpython#86276]
BPO | [42110](https://bugs.python.org/issue42110)
--- | :---
Nosy | @asvetlov, @1st1, @sourcejedi

<sup>*Note: these values reflect the state of the issue at the time it was migrated and might not reflect the current state.*</sup>

<details><summary>Show more details</summary><p>

GitHub fields:
```python
assignee = None
closed_at = None
created_at = <Date 2020-10-21.18:44:25.842>
labels = ['3.10', 'expert-asyncio']
title = 'race condition in ThreadChildWatcher (default) and MultiLoopChildWatcher'
updated_at = <Date 2020-10-22.15:09:14.856>
user = 'https://github.com/sourcejedi'
```

bugs.python.org fields:
```python
activity = <Date 2020-10-22.15:09:14.856>
actor = 'sourcejedi'
assignee = 'none'
closed = False
closed_date = None
closer = None
components = ['asyncio']
creation = <Date 2020-10-21.18:44:25.842>
creator = 'sourcejedi'
dependencies = []
files = []
hgrepos = []
issue_num = 42110
keywords = []
message_count = 3.0
messages = ['379229', '379296', '379297']
nosy_count = 3.0
nosy_names = ['asvetlov', 'yselivanov', 'sourcejedi']
pr_nums = []
priority = 'normal'
resolution = None
stage = None
status = 'open'
superseder = None
type = None
url = 'https://bugs.python.org/issue42110'
versions = ['Python 3.10']
```

</p></details>

**Comment by saito828koki** [comment]
I cannot reproduce the same bug. (Tested environment is 3.10.6)
Is this issue already fixed?

**Comment by gvanrossum** [comment]
>  Is this issue already fixed?

I don't know, the description is hard to follow, and child watchers are a minefield (e.g. https://github.com/python/cpython/issues/94597).

**Comment by kumaraditya303** [comment]
I am unable to reproduce the reported race. FWIW `PidfdChildWatcher` is now used by default which does not has any of these issue.

**Comment by gvanrossum** [comment]
But ThreadedChildWatcher is still used on macOS and other non-Linux UNIX variants. I _think_ the OP is claiming that the race condition is in subprocess.py (not the asyncio version) and due to waiting for a process in a thread that didn't create the process. Maybe we can reproduce this outside asyncio?

**Comment by gvanrossum** [comment]
This was fixed by https://github.com/python/cpython/pull/20010/files, in particular the try/except ProcessLookupError around the os.kill() call added there.
